lib/graphql/stitching/shaper.rb

```diff
@@ -0,0 +1,84 @@
+# typed: false
+# frozen_string_literal: true
+
+module GraphQL
+  module Stitching
+    class Shaper
+      def initialize(supergraph:, document:)
+        @supergraph = supergraph
+        @document = document
+      end
+
+      def perform(raw_result)
+        if ! raw_result.key?("errors") || raw_result["errors"].empty?  # TODO should do null cleanup for errors too
+          munge_entry(raw_result["data"], @document.operation.selections, @supergraph.schema.query)
+        end
+
+        # hate doing a second pass, but cannot remove _STITCH_ fields until the fragements are processed
+        clean_entry(raw_result["data"])
```

Comment:
This should also be conditional on the presence of `"data"`, which doesn't have to be there in the event of complete data loss.